File: src/main/java/net/hectus/neobb/turn/person_game/other/PTPainting.java
package net.hectus.neobb.turn.person_game.other;

import com.marcpg.libpg.util.Randomizer;
import net.hectus.neobb.buff.Buff;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.turn.default_game.other.OtherTurn;
import net.hectus.neobb.turn.person_game.categorization.BuffCategory;
import net.hectus.neobb.util.MinecraftTime;
import net.hectus.neobb.util.Modifiers;
import org.bukkit.Material;
import org.bukkit.entity.Painting;
import org.bukkit.inventory.ItemStack;
import xyz.xenondevs.invui.gui.Gui;
import xyz.xenondevs.invui.window.Window;

import java.util.List;

public class PTPainting extends OtherTurn<Painting> implements BuffCategory {
    public PTPainting(NeoPlayer player) { super(player); }
    public PTPainting(Painting data, NeoPlayer player) { super(data, data.getLocation(), player); }

    @Override
    public ItemStack item() {
        return new ItemStack(Material.PAINTING);
    }

    @Override
    public List<Buff> buffs() {
        return List.of();
    }

    @Override
    public void apply() {
        super.apply();
        switch (data.getArt()) {
            case ALBAN -> new Buff.Luck(Buff.BuffTarget.YOU, 30).apply(player);
            case AZTEC, AZTEC2 -> {
                player.inventory.removeRandom();
                player.inventory.removeRandom();
            }
            case BOMB -> player.game.players().forEach(p -> p.damage(5.0));
            case KEBAB -> player.inventory.removeRandom();
            case PLANT -> new Buff.ExtraTurn(Buff.BuffTarget.YOU, Randomizer.boolByChance(25) ? 2 : 1).apply(player);
            case WASTELAND -> player.game.players().forEach(p -> {
                p.inventory.clear();
                p.inventory.fillInRandomly();
            });
            case COURBET -> player.game.world().setTime(MinecraftTime.MIDNIGHT);
            case POOL -> player.addLuck(99);
            case SEA, BUST -> player.game.eliminatePlayer(player);
            case CREEBET -> {
                Gui.Builder.Normal gui = Gui.normal().setStructure("0 1 2 3 4 5 6 7 8");
                ItemStack[] deck = player.nextPlayer().inventory.deck();
                for (int i = 0; i < deck.length; i++)
                    gui.addIngredient((char) ('0' + i), deck[i]);

                Window.single().setTitle("Opponent Deck").setGui(gui.build()).open(player.player);
            }
            case SUNSET -> {
                List<NeoPlayer> players = player.game.players();
                double temp = players.getFirst().health();
                for (int i = 0; i < players.size() - 1; i++) {
                    players.get(i).health(players.get(i + 1).health());
                }
                players.getLast().health(temp);
            }
            case GRAHAM -> player.addModifier(Modifiers.P_NO_MOVE);
            case WANDERER -> player.game.addModifier(Modifiers.G_PERSON_NO_WIN_CONS);
            case MATCH -> player.addLuck(20);
            case SKULL_AND_ROSES -> player.game.arena.clear();
            case STAGE -> player.opponents(true).forEach(p -> {
                p.player.getLocation().getBlock().setType(Material.COBWEB);
                if (Randomizer.boolByChance(20))
                    p.game.eliminatePlayer(p);
            });
            case VOID -> {
                new Buff.ExtraTurn().apply(player);
                player.addArmor(1);
            }
            case WITHER -> player.game.eliminatePlayer(Randomizer.boolByChance(20) ? player.nextPlayer() : player);
        }
    }
}

File: src/main/kotlin/net/hectus/neobb/modes/turn/person_game/other/PTPainting.kt
package net.hectus.neobb.modes.turn.person_game.other

import com.marcpg.libpg.storing.Cord
import com.marcpg.libpg.util.MinecraftTime
import com.marcpg.libpg.util.Randomizer
import net.hectus.neobb.buff.Buff
import net.hectus.neobb.buff.ExtraTurn
import net.hectus.neobb.buff.Luck
import net.hectus.neobb.modes.turn.default_game.other.OtherTurn
import net.hectus.neobb.modes.turn.person_game.categorization.BuffCategory
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.util.Modifiers
import org.bukkit.Art
import org.bukkit.Material
import org.bukkit.entity.Painting
import org.bukkit.inventory.ItemStack
import xyz.xenondevs.invui.gui.Gui
import xyz.xenondevs.invui.window.Window

class PTPainting(data: Painting?, cord: Cord?, player: NeoPlayer?) : OtherTurn<Painting>(data, cord, player), BuffCategory {
    override fun item(): ItemStack = ItemStack(Material.PAINTING)

    override fun buffs(): List<Buff<*>> = emptyList()

    override fun apply() {
        player!! // Force non-null.
        when (data!!.art) {
            Art.ALBAN -> Luck(30).apply(player)
            Art.AZTEC, Art.AZTEC2 -> {
                player.inventory.removeRandom()
                player.inventory.removeRandom()
            }
            Art.BOMB -> player.game.players.forEach { it.damage(5.0) }
            Art.KEBAB -> player.inventory.removeRandom()
            Art.PLANT -> ExtraTurn(if (Randomizer.boolByChance(25.0)) 2 else 1).apply(player)
            Art.WASTELAND -> player.game.players.forEach { p ->
                p.inventory.clear()
                p.inventory.fillInRandomly()
            }
            Art.COURBET -> player.game.time = MinecraftTime.MIDNIGHT
            Art.POOL -> player.addLuck(99)
            Art.SEA, Art.BUST -> player.game.eliminate(player)
            Art.CREEBET -> {
                val gui = Gui.normal().setStructure("0 1 2 3 4 5 6 7 8 9")
                val deck = player.nextPlayer().inventory.deck
                for (i in deck.indices)
                    gui.addIngredient(('0'.code + i).toChar(), deck[i] ?: ItemStack.empty())

                Window.single().setTitle("Opponent Deck").setGui(gui.build()).open(player.player)
            }
            Art.SUNSET -> {
                val players = player.game.players
                val temp = players.first().health
                for (i in 0..<players.size - 1) {
                    players[i].health(players[i + 1].health)
                }
                players.last().health(temp)
            }
            Art.GRAHAM -> player.addModifier(Modifiers.Player.NO_MOVE)
            Art.WANDERER -> player.game.addModifier(Modifiers.Game.Person.NO_WIN_CONS)
            Art.MATCH -> player.addLuck(20)
            Art.SKULL_AND_ROSES -> player.game.arena.clear()
            Art.STAGE -> player.opponents().forEach { p ->
                p.location().block.type = Material.COBWEB
                if (Randomizer.boolByChance(20.0))
                    p.game.eliminate(p)
            }
            Art.VOID -> {
                ExtraTurn().apply(player)
                player.addArmor(1.0)
            }
            Art.WITHER -> player.game.eliminate(if (Randomizer.boolByChance(20.0)) player.nextPlayer() else player)
        }
    }
}
